### PR TITLE
feat: Links receive CozyClient

### PR DESCRIPTION
**Pull request by ptbrowne**
This PR supersedes https://github.com/cozy/cozy-client/pull/247.

The goal of this PR was for a link to be able to inject data in the store of CozyClient. This is necessary for the PouchLink that receives data asynchronously after a synchronization.

Previously, in Banks, the onSync handler was used to manually `setData`.
https://github.com/cozy/cozy-banks/blob/697ac7f8bb328cbc7d83f34145b811f0c7d34831/src/ducks/client/mobile.js#L80

This was done as a temporary solution. Now, this solution is backported into CozyClient. The link receives a reference to the CozyClient at registration time, saves the reference, and call `setData` when it receives new data asynchronously.

The links already received a `StackClient` at registration time but it was wrongly named `client` which caused confusion between CozyClient and CozyStackClient. To clear this confusion, this PR renames all places where a StackClient is used under the name `client` into `stackClient`. This addresses the concerns in https://github.com/cozy/cozy-client/issues/246.

## Discussions

I am not very happy with the name `setData`, I have the feeling that a link should
not directly call a method like this. I could be called `requestSetData`, or `onReceiveData`.

Should the links receive a full blown client or just a subset of its API (`getToken`, `getURL`, `requestSetData`, `refreshToken`) in order to respect the Law of Demeter ? I am not very happy that the links know that there is a stack client in the client they receive.

Should the links instantiate their own stack client when they need it instead of using the one from the cozy client ? This would make the coupling between the links and the client less strong.

### CODE REVIEW

File: packages/cozy-pouch-link/src/jsonapi.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 export const normalizeDoc = (doc, doctype) => {
   const id = doc._id || doc.id
-  return { id, _id: id, _type: doctype, ...doc }
+  const rev = doc._rev || doc.rev
```
Comment (kosssi) [inline packages/cozy-pouch-link/src/jsonapi.js]: When you update a document pouch response include `rev` but the document have already old `_rev` so it's indispensable to change it:
```suggestion
  const rev = doc.rev || doc._rev
```

File: packages/cozy-pouch-link/src/jsonapi.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
+  return { id, _id: id, rev, _rev: rev, _type: doctype, ...doc }
```
Comment (kosssi) [inline packages/cozy-pouch-link/src/jsonapi.js]: I suggeste to not create the `rev` key but use always `_rev`:
```suggestion
  const normalized = { ...doc, id, _id: id, _rev: rev, _type: doctype }
  delete normalized.rev

  return normalized
```

File: packages/cozy-pouch-link/src/jsonapi.js [hunk 3]
```diff
@@ -1,7 +1,22 @@
 export const normalizeDoc = (doc, doctype) => {
   const id = doc._id || doc.id
-  const rev = doc._rev || doc.rev
-  return { id, _id: id, rev, _rev: rev, _type: doctype, ...doc }
+
+  // PouchDB sends back .rev attribute but we do not want to
+  // keep it on the server.
+  // The rev can be higher than _rev, we keep the highest not
+  // to have any document conflict.
+  const _rev = doc.rev > doc._rev ? doc.rev : doc._rev
```
Comment (kosssi) [inline packages/cozy-pouch-link/src/jsonapi.js]: If .rev or ._rev is not defined this throw an error.

File: packages/cozy-pouch-link/src/jsonapi.js (hunk 1 above)
Comment (ptbrowne) [reply packages/cozy-pouch-link/src/jsonapi.js]: Thanks !

File: packages/cozy-pouch-link/src/jsonapi.js (hunk 2 above)
Comment (ptbrowne) [reply packages/cozy-pouch-link/src/jsonapi.js]: Done !

File: packages/cozy-stack-client/src/__tests__/setup.js [hunk 4]
```diff
@@ -35,6 +35,17 @@ expect.extend({
   }
 })
 
+beforeEach(() => {
+  jest.spyOn(global.console, 'log').mockImplementation(() => {})
+  jest.spyOn(global.console, 'info').mockImplementation(() => {})
+})
```
Comment (y-lohse) [inline packages/cozy-stack-client/src/__tests__/setup.js]: What is the reason behind this change?

Comment (ptbrowne) [reply packages/cozy-stack-client/src/__tests__/setup.js]: So many console.log polluting the test output. It's better to display them on a test by test basis, with a `console.log.mockRestore()`.

Comment (y-lohse) [reply packages/cozy-stack-client/src/__tests__/setup.js]: Meh, ok, but we should eventually review why we are writing so much to the console :)

Comment (ptbrowne) [reply packages/cozy-stack-client/src/__tests__/setup.js]: When I write to the console, it is to have useful debugging imformation. It is wrapped in a `NODE_ENV !== 'production'`.
